move_base/src/move_base.cpp

```diff
@@ -619,54 +619,59 @@ namespace move_base {
+          std::vector<geometry_msgs::PoseStamped>* temp_plan = planner_plan_;
 
-        ROS_DEBUG_NAMED("move_base_plan_thread","Generated a plan from the base_global_planner");
+          lock.lock();
+          planner_plan_ = latest_plan_;
+          latest_plan_ = temp_plan;
+          last_valid_plan_ = ros::Time::now();
+          planning_retries_ = 0;
+          new_global_plan_ = true;
 
-        //make sure we only start the controller if we still haven't reached the goal
-        if(runPlanner_)
-          state_ = CONTROLLING;
-        if(planner_frequency_ <= 0)
-          runPlanner_ = false;
-        lock.unlock();
-      }
-      //if we didn't get a plan and we are in the planning state (the robot isn't moving)
-      else if(state_==PLANNING){
-        ROS_DEBUG_NAMED("move_base_plan_thread","No Plan...");
-        ros::Time attempt_end = last_valid_plan_ + ros::Duration(planner_patience_);
-
-        //check if we've tried to make a plan for over our time limit or our maximum number of retries
-        //issue #496: we stop planning when one of the conditions is true, but if max_planning_retries_
-        //is negative (the default), it is just ignored and we have the same behavior as ever
-        lock.lock();
-        if(runPlanner_ &&
-           (ros::Time::now() > attempt_end || ++planning_retries_ > uint32_t(max_planning_retries_))){
-          //we'll move into our obstacle clearing mode
-          state_ = CLEARING;
-          publishZeroVelocity();
-          recovery_trigger_ = PLANNING_R;
+          ROS_DEBUG_NAMED("move_base_plan_thread","Generated a plan from the base_global_planner");
+
+          //make sure we only start the controller if we still haven't reached the goal
+          if(runPlanner_)
+            state_ = CONTROLLING;
+          if(planner_frequency_ <= 0)
+            runPlanner_ = false;
+          lock.unlock();
+        }
+        //if we didn't get a plan and we are in the planning state (the robot isn't moving)
+        else{
+          ROS_DEBUG_NAMED("move_base_plan_thread","No Plan...");
+          ros::Time attempt_end = last_valid_plan_ + ros::Duration(planner_patience_);
+
+          //check if we've tried to make a plan for over our time limit or our maximum number of retries
+          //issue #496: we stop planning when one of the conditions is true, but if max_planning_retries_
+          //is negative (the default), it is just ignored and we have the same behavior as ever
+          lock.lock();
+          planning_retries_++;
+          if(runPlanner_ &&
+             (ros::Time::now() > attempt_end || planning_retries_ > uint32_t(max_planning_retries_))){
+            //we'll move into our obstacle clearing mode
+            state_ = CLEARING;
+            wait_for_wake = true;
+            publishZeroVelocity();
+            recovery_trigger_ = PLANNING_R;
+          }
+
+          lock.unlock();
         }
-        lock.unlock();
       }
```

Comment:
This is where we would add "else { lock.unlock() }".

I did test this lock change, and it looks all good (also tested the whole feature in general).